### PR TITLE
RtpsRelay: bug fixes, rate limiting, performance statistics

### PR DESCRIPTION
**Pull request by jrw972**
The relay can be configured to run the relay, publish statistics, and report (print) statistics.

### CODE REVIEW

File: tools/rtpsrelay/ParticipantListener.h [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+#define RTPSRELAY_PARTICIPANT_LISTENER_H_
+
+#include "ListenerBase.h"
+#include "DomainStatisticsWriter.h"
+
+namespace RtpsRelay {
+
+class ParticipantListener : public ListenerBase {
+public:
+  ParticipantListener(DomainStatisticsWriter& stats_writer);
```
Comment (mitza-oci) [inline tools/rtpsrelay/ParticipantListener.h]:
```suggestion
  explicit ParticipantListener(DomainStatisticsWriter& stats_writer);
```

File: tools/rtpsrelay/ParticipantListener.h [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+private:
+  void on_data_available(DDS::DataReader_ptr /*reader*/) override;
```
Comment (mitza-oci) [inline tools/rtpsrelay/ParticipantListener.h]:
```suggestion
  void on_data_available(DDS::DataReader_ptr reader) override;
```

File: tools/rtpsrelay/HandlerStatisticsListener.cpp [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+                                   DDS::LENGTH_UNLIMITED,
+                                   DDS::NOT_READ_SAMPLE_STATE,
+                                   DDS::ANY_VIEW_STATE,
+                                   DDS::ALIVE_INSTANCE_STATE);
+  if (ret != DDS::RETCODE_OK) {
+    ACE_ERROR((LM_ERROR, ACE_TEXT("(%P|%t) %N:%l ERROR: HandlerStatisticsListener::on_data_available failed to read\n")));
+    return;
+  }
+
+  ACE_TCHAR timestamp[27];
```
Comment (mitza-oci) [inline tools/rtpsrelay/HandlerStatisticsListener.cpp]: Does ACE have a constant for this one?  If not we could add one in OpenDDS common libs?

File: tools/rtpsrelay/DomainStatisticsWriter.h [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+#include "lib/RelayTypeSupportImpl.h"
+#include "lib/QosIndex.h"
+#include "RelayHandler.h"
+
+namespace RtpsRelay {
+
+class DomainStatisticsWriter {
+public:
+  DomainStatisticsWriter(const RelayHandlerConfig& config)
```
Comment (mitza-oci) [inline tools/rtpsrelay/DomainStatisticsWriter.h]:
```suggestion
  explicit DomainStatisticsWriter(const RelayHandlerConfig& config)
```

File: tools/rtpsrelay/DomainStatisticsWriter.h [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    domain_statistics_.local_participants(0);
+    domain_statistics_.local_writers(0);
+    domain_statistics_.local_readers(0);
+    domain_statistics_.total_writers(0);
+    domain_statistics_.total_readers(0);
+  }
+
+  void add_local_participant()
+  {
+    domain_statistics_.local_participants() += 1;
```
Comment (mitza-oci) [inline tools/rtpsrelay/DomainStatisticsWriter.h]:
```suggestion
    ++domain_statistics_.local_participants();
```

File: tools/rtpsrelay/RtpsRelay.cpp [hunk 6]
```diff
@@ -154,137 +204,92 @@ int ACE_TMAIN(int argc, ACE_TCHAR* argv[])
...
+  if (reader_entry_ts->register_type(relay_participant, "") != DDS::RETCODE_OK) {
+    ACE_ERROR((LM_ERROR, ACE_TEXT("(%P|%t) %N:%l ERROR: failed to register ReaderEntry type\n")));
     return EXIT_FAILURE;
   }
 
-  OpenDDS::DCPS::DomainParticipantImpl* application_participant_impl =
-    dynamic_cast<OpenDDS::DCPS::DomainParticipantImpl*>(application_participant.in());
-  if (application_participant_impl == 0) {
-    ACE_ERROR((LM_ERROR, "(%P|%t) %N:%l ERROR: Failed to get application participant impl\n"));
+  DDS::Topic_var readers_topic = relay_participant->create_topic("Readers", reader_entry_ts->get_type_name(),
```
Comment (mitza-oci) [inline tools/rtpsrelay/RtpsRelay.cpp]: get_type_name allocates a string that has to be managed by String_var

File: tools/rtpsrelay/RtpsRelay.cpp [hunk 7]
```diff
@@ -359,152 +388,305 @@ int ACE_TMAIN(int argc, ACE_TCHAR* argv[])
...
+    const ACE_INET_Addr spdp_vertical_addr(port_vertical++, addr_vertical);
+    const ACE_INET_Addr sedp_vertical_addr(port_vertical++, addr_vertical);
+    const ACE_INET_Addr data_vertical_addr(port_vertical++, addr_vertical);
+
+    ACE_INET_Addr stun_addr = nic_vertical;
+    stun_addr.set_port_number(stun_port);
+
+    RelayAddresses relay_addresses {
+      addr_to_string(spdp_horizontal_addr),
+        addr_to_string(sedp_horizontal_addr),
```
Comment (mitza-oci) [inline tools/rtpsrelay/RtpsRelay.cpp]:
```suggestion
      addr_to_string(sedp_horizontal_addr),
```

Comment (mitza-oci) [reply tools/rtpsrelay/RtpsRelay.cpp]: Also take a look at the following lines, indentation seems to be off

File: tools/rtpsrelay/DomainStatisticsListener.cpp [hunk 8]
```diff
@@ -0,0 +1,51 @@
+#include "DomainStatisticsListener.h"
+
+#include "utility.h"
+
+#include "lib/RelayTypeSupportImpl.h"
+#include "lib/QosIndex.h"
+
```
Comment (mitza-oci) [inline tools/rtpsrelay/DomainStatisticsListener.cpp]:
```suggestion

#include <iostream>
```

attempt to fix some CI builds

File: tools/rtpsrelay/DomainStatisticsWriter.h (hunk 5 above)
Comment (mitza-oci) [reply tools/rtpsrelay/DomainStatisticsWriter.h]: ... and a few others in this file (`++` or `--`)

File: tools/rtpsrelay/HandlerStatisticsListener.cpp (hunk 3 above)
Comment (jrw972) [reply tools/rtpsrelay/HandlerStatisticsListener.cpp]: The code that I stole from didn't use one.  Where do you want it and what do you want it to be named?

Comment (mitza-oci) [reply tools/rtpsrelay/HandlerStatisticsListener.cpp]: `dds/DCPS/Definitions.h` ACE_timestamp_length (or similar?)

File: dds/DCPS/Definitions.h [hunk 9]
```diff
@@ -119,6 +119,11 @@ struct VarLess : public std::binary_function<V, V, bool> {
   }
 };
 
+/// Size of TCHAR buffer for use with ACE::timestamp, e.g.,
+///  ACE_TCHAR buffer[AceTimestampSize];
+///  ACE::timestamp(buffer, sizeof(buffer) / sizeof(ACE_TCHAR));
```
Comment (mitza-oci) [inline dds/DCPS/Definitions.h]:
```suggestion
///  ACE::timestamp(buffer, AceTimestampSize);
```

File: docs/guidelines.md [hunk 10]
```diff
@@ -253,6 +253,7 @@ later, `nullptr` should be used instead. `NULL` should never be used.
   - `methodsAreCamelCaseWithInitialLower` OR
     `methods_are_lower_case_with_underscores`
   - `member_data_use_underscores_and_end_with_an_underscore_`
+  - `ThisIsAGlobalConstant`
```
Comment (mitza-oci) [inline docs/guidelines.md]:
```suggestion
  - `ThisIsANamespaceScopedOrStaticClassMemberConstant`
```

**Pull request merged**